src/bindings/python/flux/job/output.py

```diff
@@ -0,0 +1,782 @@
+    Attributes:
+        stdout (str): stdout from all tasks
+        stderr (str): stderr from all tasks
+        log (str): log messages
+    """
+
+    stdout: str
+    stderr: str
+    log: str
+
+
+def _parse_output_eventlog_entry(entry, labelio=False):
+    """
+    Parse a single output eventlog entry, returning an object of the
+    appropriate type: OutputEvent, LogEvent, OutputHeaderEvent,
+    RedirectEvent or JobExceptionEvent.
+    """
+    if entry is None:
+        return None
+    if isinstance(entry, EventLogEvent):
+        event = entry
+    else:
+        event = EventLogEvent(entry)
+    name = event.name
+    if name == "data":
+        return OutputEvent(event, labelio)
+    elif name == "log":
+        return LogEvent(event)
+    elif name == "header":
+        return OutputHeaderEvent(event)
+    elif name == "redirect":
+        return RedirectEvent(event)
+    elif name == "exception":
+        # Not technically an output event, but handle anyway
+        return JobExceptionEvent(event)
+    raise ValueError(f"event {name} is not a supported output event type")
+
+
+def _output_eventlog_entry_decode(
+    entry, stream_dict, tasks, labelio=False, log_stderr_level=LOG_TRACE
+):
+    """
+    Decode RFC 24 output eventlog entry ``entry``, appending the result to
+    the stream_dict[stream_name] list if the entry is in the set set of
+    requested ``tasks``. If stream_dict[stream_name] does not exist, the
+    entry will be created.
+    """
+    if not isinstance(tasks, Taskset):
+        raise ValueError("tasks argument must be a Taskset, got " + type(tasks))
+    event = _parse_output_eventlog_entry(entry, labelio)
+
+    #  Determine stream name of this event:
+    stream = None
+    if event.name == "data":
+        if event.rank.intersect(tasks):
+            stream = event.stream
+    elif event.name == "log":
+        if event.level <= log_stderr_level:
+            stream = "stderr"
+        else:
+            stream = "log"
+    elif event.name in "exception":
+        if event.rank.intersect(tasks):
+            stream = "stderr"
+    elif event.name == "exception":
+        stream = "stderr"
+    else:
+        return
+
+    data = event.render()
+    if data:
+        stream_dict.setdefault(stream, []).append(data)
+
+
+def _parse_output_eventlog(
+    eventlog, tasks="*", labelio=False, log_stderr_level=LOG_TRACE
+):
+    """
+    Given an eventlog, return a JobOutput tuple with stdout, stderr,
+    and log streams broken out.
+    """
+    stream_dict = {"stdout": [], "stderr": [], "log": []}
+    tasks = Taskset(tasks)
+
+    for line in eventlog.splitlines():
+        _output_eventlog_entry_decode(
+            line, stream_dict, tasks, labelio, log_stderr_level
+        )
+
+    # Join lines and return result
+    results = ["".join(stream_dict.get(k)) for k in ("stdout", "stderr", "log")]
+    return JobOutput(*results)
+
+
+def wait_for_output(flux_handle, jobid):
```

Comment:
Hm, I'm not sure there's a use case for this function outside of the one place it is used (earlier iterations used it elsewhere), so maybe I'll just remove it and move the code into the function.

> I think the wait is what screwed me up, since that typically suggests "wait for job to finish".

We also have `future_wait_for()` which waits for a future to be ready. That is more along the lines of what I was thinking here -- waiting for output to be "ready" to read. (i.e. similar to the kvs WAITCREATE flag). So I don't think we can say that `wait` shall only mean "wait for job to finish" in Flux.

Reply:
Fair enough, dunno why my brain thought `wait_for_output()` would wait for the job to finish first.  Perhaps it's just me.

Reply:
Yeah, I can understand. Removed the confusing function since I'm not sure there's not much of a real use case.